### PR TITLE
Add an archive descriptor

### PR DESCRIPTION
**Pull request by nex3**

### CODE REVIEW

File: lib/src/archive_descriptor.dart [hunk 1]
```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
```
Comment (srawlins) [inline lib/src/archive_descriptor.dart]: 2019?

File: lib/src/archive_descriptor.dart [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      }
+    })))
+        .expand((files) => files);
+  }
+
+  ArchiveDescriptor(String name, Iterable<Descriptor> contents)
+      : contents = List.unmodifiable(contents),
+        super(name);
+
+  Future create([String parent]) async {
```
Comment (srawlins) [inline lib/src/archive_descriptor.dart:71]: Should this be typed as `Future<void>`, for understandability?

File: lib/src/utils.dart [hunk 3]
```diff
@@ -67,7 +89,7 @@ bool matchesAll(Pattern pattern, String string) =>
 
 /// Like [Future.wait] with `eagerError: true`, but reports errors after the
 /// first using [registerException] rather than silently ignoring them.
-Future waitAndReportErrors(Iterable<Future> futures) {
+Future<List<T>> waitAndReportErrors<T>(Iterable<Future<T>> futures) {
```
Comment (srawlins) [inline lib/src/utils.dart:92]: Very nice.

**Review by srawlins — APPROVED** [review]
Looks good!

File: lib/src/archive_descriptor.dart (hunk 1 above)
Comment (nex3) [reply lib/src/archive_descriptor.dart]: Done.

File: lib/src/archive_descriptor.dart (hunk 2 above)
Comment (nex3) [reply lib/src/archive_descriptor.dart:71]: I was matching the superclass's signature. Probably the whole package should be converted to `Future<void>`, but I don't think this is the PR for it.

**Pull request merged**